### PR TITLE
Add std::variant backport as c10::variant

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26837 [BC-breaking] Use torch::variant for enums in C++ API
* **#26836 Add std::variant backport as c10::variant**

Differential Revision: [D17649064](https://our.internmc.facebook.com/intern/diff/D17649064)

### CODE REVIEW

**Comment by ezyang** [comment]
I don't think variant.h should live in the torch. c10/util seems like a much more appropriate spot.

**Comment by yf225** [comment]
@ezyang I was a bit concerned about moving it to c10 because `variant` is not used in c10/ATen at the moment, but moving it seems to make code organization better.

File: c10/util/variant.h [hunk 1]
```diff
@@ -0,0 +1,2824 @@
...
+  template <typename... Args>                     \
+  inline explicit constexpr type(Args &&... args) \
+      : base(lib::forward<Args>(args)...) {}
+#endif
+
+    template <typename Traits, Trait = Traits::destructible_trait>
+    class destructor;
+
+#define MPARK_VARIANT_DESTRUCTOR(destructible_trait, definition, destroy) \
+  template <typename... Ts>
```
Comment (ezyang) [inline c10/util/variant.h]: The way we have historically done these ports is to literally rename the header into our namespace, without leaving behind the old namespace. The reason for this is if libtorch is used in a library that itself is using mpark variant, if you put the implementation in the old mpark namespace, you will have two incompatible definitions of the same code, and compilation will fail.

**Review by ezyang — CHANGES_REQUESTED** [review]
variant.h needs to be renamespaced directly into c10

File: c10/util/variant.h (hunk 1 above)
Comment (yf225) [reply c10/util/variant.h]: Sounds great, I updated it accordingly.

File: c10/util/variant.h [hunk 2]
```diff
@@ -2817,7 +2817,7 @@ namespace std {
 
 namespace c10 {
 
-using namespace mpark;
+using namespace c10;
```
Comment (ezyang) [inline c10/util/variant.h]: Drop this now

**Review by ezyang — APPROVED** [review]
The tests are in the wrong place but I'll go ahead and approve this.

**Comment by yf225** [comment]
I moved the test to `aten/src/ATen/test/variant_test.cpp`.

**Comment by kostmo** [comment]
This appears to have broken `caffe2_py2_clang3_8_ubuntu16_04_build` with the following error:
```
variant_test.cpp:15:23: error: default initialization of an object of const type 'const enumtype::Enum3' without a user-provided default constructor
```
Reverting.

**Pull request closed without merging**